test/astarte/client/appengine_test.exs

```diff
@@ -0,0 +1,93 @@
+#
+# This file is part of Astarte.
+#
+# Copyright 2021 SECO Mind
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+defmodule Astarte.Client.AppEngineTest do
+  use ExUnit.Case
+  doctest Astarte.Client.AppEngine
+
+  alias Astarte.Client.AppEngine
+
+  @base_url "https://base-url.com"
+  @realm_name "realm_name"
+  @jwt "notarealjwt"
+  @valid_private_key X509.PrivateKey.new_ec(:secp256r1) |> X509.PrivateKey.to_pem()
+  @valid_private_key_algorithm "ES256"
+  @invalid_private_key "notaprivatekey"
+
+  describe "new/3" do
+    test "new/3 with jwt creates client" do
+      assert {:ok, %AppEngine{http_client: http_client}} =
+               AppEngine.new(@base_url, @realm_name, jwt: @jwt)
+
+      with middleware = Tesla.Client.middleware(http_client),
+           appengine_url = get_appengine_url(middleware),
+           headers = get_headers(middleware),
+           jwt = get_jwt(headers) do
+        assert appengine_url == @base_url <> "/appengine/v1/" <> @realm_name
+        assert jwt == @jwt
+      end
```

Comment:
Tests should not leak internal details about the specific library we're using, remove these assertions. You should just verify that `new` returns `{:ok, %Appengine{}}`, since right now the client is basically a black box you pass as first parameter. This guarantees that if we were to change internal implementation details, e.g. the http client, the tests would remain valid.

If you want to test that the client actually uses the base url and the headers to send the request, do a separate test like
```elixir
test "the client uses the information to perform the request" do
  ...
end
```
and use `Tesla.mock`* to verify that the request is sent to the correct base URL with the correct headers. When you verify headers, use `in` to assert in the headers list (e.g. `assert {"Authorization", "Bearer# #{jwt} in headers`) instead of relying on the order or explicitly extracting from the list like you were doing here.

*This actually is, again, coupling tests with the specific HTTP library, but it's ok since we can always replace this with Mox in a second pass.